jc/parsers/find.py

```diff
@@ -0,0 +1,121 @@
+"""jc - JSON Convert `find` command output parser
+
+Usage (cli):
+
+    $ find | jc --find
+
+Usage (module):
+
+    import jc
+    result = jc.parse('find', find_command_output)
+
+Schema:
+
+    [
+      {
+        "directory path":           string,
+        "file or empty directory:   string
+      }
+    ]
+
+Examples:
+
+    $ find | jc --find -p
+    [
+        {
+          "directory path": "./directory"
+          "file or empty directory": "filename"
+        },
+        {
+          "directory path": "./anotherdirectory"
+          "file or empty directory": "anotherfile"
+        },
+        ...
+    ]
+
+    $ find | jc --find -p -r
+    {
+      "./templates/readme_template",
+      "./templates/manpage_template",
+      "./.github/workflows/pythonapp.yml",
+      ...
+    }
+"""
+import jc.utils
+
+
+class info():
+    """Provides parser metadata (version, author, etc.)"""
+    version = '1.0'
+    description = '`find` command parser'
+    author = 'Solomon Leang'
+    author_email = 'solomonleang@gmail.com'
+    compatible = ['linux']
+    tags = ['command']
+
+
+__version__ = info.version
+
+
+def _process(proc_data):
+    """
+    Final processing to conform to the schema.
+
+    Parameters:
+
+        proc_data:   (Dictionary) raw structured data to process
+
+    Returns:
+
+        List of Dictionaries. Structured data to conform to the schema.
+    """
+
+    processed = []
+    for i in proc_data:
+        try:
+            temp = str(i).rindex("/")
+        except ValueError:
+            pass
+        proc_line = {
+            'directory path': i[:temp + 1],
+            'file or empty directory': i[temp + 1:]
+        }
+        
+        processed.append(proc_line)
+    return processed
+    
+
+
+def parse(data, raw=False, quiet=False):
+    """
+    Main text parsing function
+
+    Parameters:
+
+        data:        (string)  text data to parse
+        raw:         (boolean) unprocessed output if True
+        quiet:       (boolean) suppress warning messages if True
+
+    Returns:
+
+        Dictionary of raw structured data or
+        List of Dictionaries of processed structured data
+    """
+    jc.utils.compatibility(__name__, info.compatible, quiet)
+    jc.utils.input_type_check(data)
+
+    raw_output = []
+    
+    if jc.utils.has_data(data):
+        previous = ""
+        new_line = data.split("\n")
+        for index in new_line:
+            if (index[:len(previous)] != previous):
+                raw_output.append(previous)
+            previous = index
+        raw_output.append(previous)
```

Comment:
Not sure I understand what this is doing. Is this removing duplicates?

Reply:
Essentially what it is doing is checking to see if the previous path is contained in the current index (path). This would ensure every directory is accounted for and all files (or empty directories) would get added into the list of raw_output. The last statement there ensures the last file gets added to the list. Let me know if this makes more sense!

Reply:
I don't understand why, though. If the `find` output is just a newline terminated list of paths like this:
```
.
./
./jc
./jc/tests
./jc/tests/test_find.py
./jc/tests/test_history.py
./jc/tests/test_hosts.py
./jc/anotherdirectory
```
Wouldn't just running it through `splitlines()` to turn it into a list do the same thing? Or are you checking for paths with newline characters in them or something? I'm probably missing something simple. :)

Reply:
Ah I see what you mean. I am removing the lines where it is only the directory being listed (and that the directory is not empty).
So in the example you provided, it would skip these:
```
.
./
./jc
./jc/tests
```
because they aren't files and the directory is not empty.

Thats why the output would only be something like below (also assuming anotherdirectory is empty in this case!)
```
[{"directory path": "./jc/tests/","file or empty directory": "test_find.py"},
{"directory path": "./jc/tests/","file or empty directory": "test_history.py"},
{"directory path": "./jc/tests/","file or empty directory": "test_hosts.py"},
{"directory path": "./jc/","file or empty directory": "anotherdirectory"}]
``` 

However, I can include these 4 lines instead and change the schema so that it is just `path` and `node` where `node` can be a nonempty directory as well; or, would skipping these lines still be acceptable?

Reply:
Interesting. I see how skipping those lines makes sense, but on the other hand, they are part of the find output so it seems we should just leave them in. The downstream application can decide to ignore those lines if they want with their own logic.